napari/components/viewer_model.py

```diff
@@ -1000,6 +1000,73 @@ def add_vectors(
         self.add_layer(layer)
         return layer
 
+    def _add_layer_from_data(
+        self, data, meta: dict = None, layer_type: str = 'image'
+    ):
+        """Add arbitrary layer data to the viewer.
+
+        Primarily intended for usage by reader plugin hooks.
+
+        Parameters
+        ----------
+        data : Any
+            Data in a format that is valid for the corresponding `add_*` method
+            of the specified ``layer_type``.
+        meta : dict, optional
+            Dict of keyword arguments that will be passed to the corresponding
+            `add_*` method.  MUST NOT contain any keyword arguments that are
+            not valid for the corresponding method.
+        layer_type : str
+            Type of layer to add.  MUST have a corresponding add_* method on
+            on the viewer instance.
+
+        Raises
+        ------
+        ValueError
+            If ``layer_type`` is not one of the recognized layer types.
+        TypeError
+            If any keyword arguments in ``meta`` are unexpected for the
+            corresponding `add_*` method for this layer_type.
+
+        Examples
+        --------
+        A typical use case might be to upack a tuple of layer data with a
+        specified layer_type.
+
+        >>> viewer = napari.Viewer()
+        >>> data = (
+        ...     np.random.random((10, 2)) * 20,
+        ...     {'face_color': 'blue'},
+        ...     'points',
+        ... )
+        >>> viewer._add_layer_from_data(*data)
+
+        """
+
+        if layer_type not in layers.names:
+            raise ValueError(
+                f"Unrecognized layer_type: '{layer_type}'. "
+                f"Must be one of: {layers.names}."
+            )
+
+        try:
+            add_method = getattr(self, 'add_' + layer_type)
```

Comment:
It might be nice to use `layer_type.lower()` in case people use a different case. I could see getting confused since the class names are capitalized.

Reply:
Although, I guess they would have received the very clear `ValueError` above. I'm happy with whatever you think is best!